### Library Imports

In [23]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

### Exploratory Data Analysis

In [4]:
df = pd.read_csv('ProjectDataset.csv')
df.head()

state      playerid       wagerid             event_start placed_date  \
0  State1  3.065121e+07  1.693004e+06  2021-04-28 00:30:00+00  2021-04-27   
1  State1  2.223717e+07  1.696371e+06  2021-04-28 01:45:00+00  2021-04-27   
2  State1  2.223717e+07  1.696371e+06  2021-04-28 01:45:00+00  2021-04-27   
3  State1  2.223717e+07  1.696371e+06  2021-04-28 01:45:00+00  2021-04-27   
4  State1  2.223717e+07  1.696371e+06  2021-04-28 01:45:00+00  2021-04-27   

  settled_date sportname  bet_type result  net_stake   ggr legresult  \
0   2021-04-27       nhl  straight    won       6.64 -4.96       won   
1   2021-04-27       nba    parlay   lost       5.00  5.00       won   
2   2021-04-27       nba    parlay   lost       5.00  5.00      lost   
3   2021-04-27       nba    parlay   lost       5.00  5.00      lost   
4   2021-04-27       nba    parlay   lost       5.00  5.00      lost   

   decimalodds  
0      1.74627  
1      1.78125  
2      1.86207  
3      1.74627  
4      1.78125

In [6]:
# rows and columns count

df.shape

(4174500, 13)

In [7]:
# names of columns

df.columns.values

array(['state', 'playerid', 'wagerid', 'event_start', 'placed_date',
       'settled_date', 'sportname', 'bet_type', 'result', 'net_stake',
       'ggr', 'legresult', 'decimalodds'], dtype=object)

In [10]:
# count of unique values in each column

df.nunique()

state                 3
playerid          36387
wagerid         2402300
event_start       11991
placed_date         366
settled_date        367
sportname             7
bet_type              2
result                2
net_stake         31701
ggr               96979
legresult             5
decimalodds       18694
dtype: int64

In [19]:
# where do null values exist in the dataframe

df.isna().sum()

state              0
playerid           0
wagerid            0
event_start        0
placed_date        0
settled_date       0
sportname          0
bet_type           0
result             0
net_stake          0
ggr                0
legresult          0
decimalodds     2472
dtype: int64

There are 4174500 rows and 13 columns in the dataset. There are 2472 null values, all of which are in the 'decimalodds' columns.

In [41]:
df.describe()

playerid       wagerid     net_stake           ggr   decimalodds
count  4.174500e+06  4.174500e+06  4.174500e+06  4.174500e+06  4.172028e+06
mean   2.273927e+07  4.754173e+07  2.361655e+01  2.451643e+00  3.984459e+00
std    1.033895e+07  3.675171e+07  8.189246e+01  8.532259e+01  1.602930e+01
min    5.971640e+04  1.691622e+06  4.001000e+00 -2.100000e+04  1.000100e+00
25%    1.750210e+07  2.296253e+07  5.000000e+00  0.000000e+00  1.649350e+00
50%    2.379874e+07  3.859404e+07  1.000000e+01  6.000000e+00  1.909090e+00
75%    3.064089e+07  5.861577e+07  2.000000e+01  1.200000e+01  2.200000e+00
max    4.297671e+07  1.779192e+08  1.967200e+04  1.967200e+04  5.001000e+03

ggr values on average are slightly higher than the net_stake values. How to increase the margin?

### Examining Null Values

In [21]:
null_data = df[df.isnull().any(axis=1)]
null_data.head()

state      playerid       wagerid             event_start  \
936440  State1  2.227482e+07  4.895481e+07  2021-03-29 23:25:00+00   
936441  State1  2.227482e+07  4.895481e+07  2021-03-30 02:02:00+00   
936833  State1  2.886949e+07  4.911740e+07  2021-03-30 23:25:00+00   
936834  State1  2.886949e+07  4.911740e+07  2021-03-31 02:07:00+00   
936850  State1  3.091458e+07  4.912196e+07  2021-03-31 02:07:00+00   

       placed_date settled_date           sportname bet_type result  \
936440  2021-03-29   2021-03-30  college basketball   parlay    won   
936441  2021-03-29   2021-03-30  college basketball   parlay    won   
936833  2021-03-30   2021-03-31  college basketball   parlay   lost   
936834  2021-03-30   2021-03-31  college basketball   parlay   lost   
936850  2021-03-30   2021-03-31  college basketball   parlay   lost   

        net_stake   ggr legresult  decimalodds  
936440       5.00 -6.25       won          NaN  
936441       5.00 -6.25       won          NaN  
936833       5.00  5.00       won          NaN  
936834       5.00  5.00      lost          NaN  
936850       4.84  4.84      lost          NaN

In [30]:
null_data.shape

(2472, 13)

In [34]:
null_data.nunique()

state             3
playerid        288
wagerid         928
event_start     394
placed_date      88
settled_date     84
sportname         3
bet_type          1
result            2
net_stake       191
ggr             248
legresult         3
decimalodds       0
dtype: int64

In [35]:
null_data.sportname.value_counts()

nba                   1823
college basketball     638
college football        11
Name: sportname, dtype: int64

Rows with null values seem to be random and shouldn't be counted as outlier data points so no reason to drop them.

### Examining Individual Columns

In [26]:
# how many states are included in dataframe

df.state.value_counts(normalize=True)

State2    0.400008
State3    0.330195
State1    0.269797
Name: state, dtype: float64

In [25]:
# which sports are included in the dataframe

df.sportname.value_counts(normalize=True)

nfl                   0.329091
nba                   0.325649
college basketball    0.119011
mlb                   0.113918
college football      0.062496
nhl                   0.044950
champions league      0.004885
Name: sportname, dtype: float64

In [27]:
# what kinds of bets were placed

df.bet_type.value_counts(normalize=True)

parlay      0.589196
straight    0.410804
Name: bet_type, dtype: float64

In [28]:
# wins vs losses

df.result.value_counts(normalize=True)

lost    0.732141
won     0.267859
Name: result, dtype: float64

In [29]:
# 

df.legresult.value_counts(normalize=True)

lost       0.505631
won        0.488063
void       0.005670
open       0.000626
unknown    0.000010
Name: legresult, dtype: float64

In [38]:
# which kinds of bets produced the highest ggr

df.ggr.sort_values(ascending=False)

3330266    19672.00
102129     12000.00
565327     12000.00
3224061    10000.00
2122683     8743.32
             ...   
2532086   -10000.00
3737475   -10909.09
3329787   -14297.50
1057409   -19620.00
557657    -21000.00
Name: ggr, Length: 4174500, dtype: float64

In [48]:
df['net_stake'].groupby(df['playerid']).value_counts()

playerid      net_stake 
5.971640e+04  5.000000      141
              10.000000      66
              6.666666       56
              7.500000       46
              25.000000      46
                           ... 
4.297319e+07  200.000000      1
4.297671e+07  25.000000       2
              10.000000       1
              20.000000       1
              30.000000       1
Name: net_stake, Length: 567135, dtype: int64

### Visualizations

### Things to look at

In [ ]:
# histogram breaking down the number of bets placed by people, how to count that?
# which sports have the most returning users, what is the amount that they are betting, wins/losses?
# 